RuminantFarmSystems/RuFaS
# [EEE] Fixed bug related to time cutoff in Emissions estimator

Fixes a bug in how crop yields, fertilizer applications, and manure stuff are filtered when processing emissions.

### Context

Issue(s) closed by this pull request: closes #1899

      
How big are the changes in the PR? Would you nominate it for a major version upgrade?

### What

Changes how crop harvests, manure requests and applications, and fertilizer applications are filtered when collecting these events from the last 365 days of the simulation.

### Why

Previously, the day and year cutoffs were checked separately and if an event was not "after" the cutoff for one then it would not be considered as part of the final 365 days of the year.

### How

The date cutoff and the dates of crop harvests, etc. are converted to `datetime.date`s with `Time` utility function for doing this. Then they are compared to see if the date passes the cutoff.

Note that the Mypy count is increased because the return type of `Time.convert_year_jday_to_date` is incorrect.

### Test plan

1. Switch to `FARMES_fakefarm1` branch.
2. Merge `fix-emissions-estimator-time-error` branch in.
3. Modify the `farm_es_metadata.json` to use `default_feed.json` instead of `feed.json` for the "feed", to use the new purchased feeds emissions files from 2024.
4. Run `python main.py` with the filter provided in the issue.
5. After running the simulation, modify the end date in `config.json` to be December 24, 2023.
6. Run `python main.py` again.
7. Compare the homegrown feeds emissions outputs from the two runs. All outputs related to crops should be the same.
